modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/introspection/validation/ExclusiveParameterModelValidator.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.introspection.validation;
+
+import static java.lang.String.format;
+import static org.mule.runtime.module.extension.internal.util.IntrospectionUtils.getFieldMetadataType;
+import org.mule.metadata.api.ClassTypeLoader;
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.metadata.api.model.SimpleType;
+import org.mule.metadata.api.visitor.MetadataTypeVisitor;
+import org.mule.runtime.extension.api.annotation.ExclusiveOptionals;
+import org.mule.runtime.extension.api.exception.IllegalModelDefinitionException;
+import org.mule.runtime.extension.api.introspection.EnrichableModel;
+import org.mule.runtime.extension.api.introspection.ExtensionModel;
+import org.mule.runtime.extension.api.introspection.config.ConfigurationModel;
+import org.mule.runtime.extension.api.introspection.connection.ConnectionProviderModel;
+import org.mule.runtime.extension.api.introspection.declaration.type.ExtensionsTypeLoaderFactory;
+import org.mule.runtime.extension.api.introspection.operation.OperationModel;
+import org.mule.runtime.extension.api.introspection.source.SourceModel;
+import org.mule.runtime.module.extension.internal.exception.IllegalParameterModelDefinitionException;
+import org.mule.runtime.module.extension.internal.introspection.ParameterGroup;
+import org.mule.runtime.module.extension.internal.model.property.ParameterGroupModelProperty;
+import org.mule.runtime.module.extension.internal.util.IdempotentExtensionWalker;
+
+import java.util.Optional;
+
+/**
+ * This validator makes sure that all the {@link OperationModel}, {@link SourceModel}, {@link ConnectionProviderModel}
+ * and {@link ConfigurationModel} which contains any {@link ParameterGroup} using exclusion complies the following condition:
+ * <p>
+ * The class of the {@link ParameterGroup} doesn't contain any nested {@link ParameterGroup} or any other parameter of a complex type.
+ *
+ * @since 4.0
+ */
+public final class ExclusiveParameterModelValidator implements ModelValidator
+{
+
+    private ClassTypeLoader typeLoader = ExtensionsTypeLoaderFactory.getDefault().createTypeLoader();
+    private final static String SOURCE = "source";
```

Comment:
looks like you're using this constants just to construct the message. It's the same hack you did on the parsers. Instead, add an utils in IntrospectionUtils  which given a model returns the proper String